### PR TITLE
[MNG-7374] Mutating RelocatedArtifact does not retain type

### PR DESCRIPTION
**Pull request by michael-o**
This closes #641

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [x] Make sure there is a [JIRA issue](https://issues.apache.org/jira/browse/MNG) filed 
       for the change (usually before you start working on it).  Trivial changes like typos do not 
       require a JIRA issue.  Your pull request should address just this issue, without 
       pulling in other changes.
 - [x] Each commit in the pull request should have a meaningful subject line and body.
 - [x] Format the pull request title like `[MNG-XXX] - Fixes bug in ApproximateQuantiles`,
       where you replace `MNG-XXX` with the appropriate JIRA issue. Best practice
       is to use the JIRA issue title in the pull request title and in the first line of the 
       commit message.
 - [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [x] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.
 - [x] You have run the [Core IT][core-its] successfully.

If your pull request is about ~20 lines of code you don't need to sign an
[Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf) if you are unsure
please ask on the developers list.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [ ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

 - [ ] In any other case, please file an [Apache Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

[core-its]: https://maven.apache.org/core-its/core-it-suite/

### CODE REVIEW

File: maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/RelocatedArtifact.java [hunk 1]
```diff
@@ -86,6 +86,39 @@ public String getVersion()
         }
     }
 
+    @Override
+    public Artifact setVersion( String version )
+    {
+         String current = getVersion();
+         if ( current.equals( version ) || ( version == null && current.length() <= 0 ) )
```
Comment (Marcono1234) [inline maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/RelocatedArtifact.java:93]: Could this be simplified?
```suggestion
         if ( current.equals( version ) || ( version == null && current.isEmpty() ) )
```

Comment (michael-o) [reply maven-resolver-provider/src/main/java/org/apache/maven/repository/internal/RelocatedArtifact.java:93]: While this is true, I wanted to retain the same code block as in `AbstractArtifact` for consistency reasons

**Comment by hboutemy** [comment]
@michael-o instead of copy/pasting the 3 methods, I think we could change "private Artifact newInstance" (from AbstractArtifact) to protected, then simply override this method

**Comment by michael-o** [comment]
> @michael-o instead of copy/pasting the 3 methods, I think we could change "private Artifact newInstance" (from AbstractArtifact) to protected, then simply override this method

While you are absolutely right, this is not the target of this ticket OR I need to broaden the ticket and solve this for all derved types. I absolutely do not like the fact that `AbstractArtifact` lacks abstraction and uses `DefaultArtifact` internally.

Maybe: Mutating derived Artifact does not retain type

**Comment by hboutemy** [comment]
no, I'm not trying to automagically retain the original type: I'm just trying to use the (hand-written) factory to remove code duplication

digging a little bit more, it requires a maven-resolver release, which makes the update harder from a release perspective

**Comment by michael-o** [comment]
> no, I'm not trying to automagically retain the original type: I'm just trying to use the (hand-written) factory to remove code duplication

Agree, suboptimal.

> digging a little bit more, it requires a maven-resolver release, which makes the update harder from a release perspective

I guess we can make that for 2.0.0 happen. Maybe 1.8.0, unsure. @cstamas

**Comment by michael-o** [comment]
@hboutemy Any objections to merge this one and resolve in general with new issues?

**Comment by michael-o** [comment]
https://issues.apache.org/jira/browse/MRESOLVER-233

**Comment by hboutemy** [comment]
@michael-o no problem to do the work in 2 steps, sure

**Comment by jira-importer** [comment]
Resolve #8063

**Pull request merged**